git/archive.go

```diff
@@ -0,0 +1,260 @@
+			switch e.Mode {
+			case ModeBlob:
+				hdr.Mode = 0644
+			case ModeExec:
+				hdr.Mode = 0755
+			default:
+				hdr.Mode = 0644
+			}
+
+			if err := tw.WriteHeader(hdr); err != nil {
+				return err
+			}
+			if _, err := tw.Write(o.GetContent()); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func createZipArchive(c *Client, opts ArchiveOptions, sha Sha1, entries []*IndexEntry) error {
+	fileOutput := os.Stdout
+
+	// If the output file is not empty use it instead of stdout
+	if opts.OutputFile != "" {
+		if file, err := os.Create(opts.OutputFile); err == nil {
+			fileOutput = file
+			defer file.Close()
+		} else {
+			return err
+		}
+	}
+
+	mtime := time.Now()
+
+	// If the sha is a tree we set the file modification time to the current time
+	// otherwise we must use the commit time.
+	if sha.Type(c) != "tree" {
+		if t, err := CommitID(sha).GetDate(c); err == nil {
+			mtime = t
+		}
+	}
+
+	zw := zip.NewWriter(fileOutput)
+	defer zw.Close()
+
+	// set the compression level
+	zw.RegisterCompressor(zip.Deflate, func(out io.Writer) (io.WriteCloser, error) {
+		return flate.NewWriter(out, opts.CompressionLevel)
+	})
+
+	// Set the commit sha as the zip comment.
+	zw.SetComment(sha.String())
+
+	for _, e := range entries {
+		o, err := c.GetObject(e.Sha1)
+		if err != nil {
+			return err
+		}
+		if obj, ok := o.(GitBlobObject); ok {
+			hdr := &zip.FileHeader{
+				Name:     opts.BasePrefix + e.PathName.String(),
+				Modified: mtime,
+				Method:   zip.Deflate,
+			}
+			f, err := zw.CreateHeader(hdr)
+
+			if err != nil {
+				return nil
+			}
+
+			if _, err := f.Write(obj.GetContent()); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// Return the list of supported archive file format
+func ArchiveFormatList() []string {
+	formats := make([]string, 0, len(supportedArchiveFormats))
+	for key := range supportedArchiveFormats {
+		formats = append(formats, key)
+	}
+	return formats
+}
+
+func Archive(c *Client, opts ArchiveOptions, arg string) error {
+	format := archiveTar
+
+	//
+	treeish, err := RevParseTreeish(c, &RevParseOptions{}, arg)
+	if err != nil {
+		return err
+	}
+
+	// find the format output desired
+	if opts.ArchiveFormat != "" {
+		format, err = findOutputFileFormat(opts.ArchiveFormat)
+		if err != nil {
+			return fmt.Errorf("Unknow archive format '%s'", opts.ArchiveFormat)
+		}
+	} else if opts.OutputFile != "" {
+		// if the output file is not empty try to find
+		// the archive format from the file extension.
+		format, _ = findOutputFileFormat(opts.OutputFile)
+	}
+
+	// commit hash
+	var sha1 Sha1
+
+	// if the input it's a tag we must resolve it to a commit
+	if h, ok := treeish.(Ref); ok {
+		refspec := RefSpec(h.Name)
+		if c, err := refspec.CommitID(c); err != nil {
+			return err
+		} else {
+			treeish = c
+		}
+	}
+
+	if s, ok := treeish.(CommitID); ok {
+		sha1 = Sha1(s)
+	} else {
+		return fmt.Errorf("Can't convert treeish to commit id")
+	}
```

Comment:
I think you're going to end up with a lot of cases/exceptions here. You probably just want to check if the type can be converted to a Commitish to get the CommitID with something like:

```
commitish, ok := treeish.(Commitish)
if ok {
      cid, err := commitish.CommitID(c)
      if err != nil {
            return err
      }
      sha1 = Sha1(cid)
}
```

but if I understand what you're doing correctly you're just using the sha1 to get the mtime, so you might also want to lift the mtime conversion all the way up to the exported Archive function and pass it down instead of passing the Sha1 down and duplicating it for zip/tar.

Reply:
It's not just to get the mtime but to set the sha1 value as a comment if the output is zip or as a pax object if it's tar

Reply:
Ah, okay, in that case you still need to pass it, but you'll still make your life easier by lifting it up a level and checking if it implements the Conmitish interface rather than trying to exhaustively check every possible Treeish implementer type

Reply:
Its updated now :-)

Reply:
You still have the tree.(Ref) above. Ref implements Commitish, so it's redundant

Reply:
Done!

Reply:
Thanks.. it all looks much cleaner now, I think.